core/coreapi/routing.go

```diff
@@ -0,0 +1,70 @@
+package coreapi
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+
+	"github.com/ipfs/go-path"
+	"github.com/libp2p/go-libp2p-core/routing"
+	peer "github.com/libp2p/go-libp2p/core/peer"
+)
+
+type RoutingAPI CoreAPI
+
+func (r *RoutingAPI) Get(ctx context.Context, key string) ([]byte, error) {
+	dhtKey, err := escapeDhtKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	ctx, events := routing.RegisterForQueryEvents(ctx)
+
+	var getErr error
+
+	go func() {
+		defer cancel()
+		var val []byte
+		val, getErr = r.routing.GetValue(ctx, dhtKey)
+		if getErr != nil {
+			routing.PublishQueryEvent(ctx, &routing.QueryEvent{
+				Type:  routing.QueryError,
+				Extra: getErr.Error(),
+			})
+		} else {
+			routing.PublishQueryEvent(ctx, &routing.QueryEvent{
+				Type:  routing.Value,
+				Extra: base64.StdEncoding.EncodeToString(val),
+			})
+		}
+	}()
+
+	for e := range events {
+		if e.Type == routing.Value {
+			return []byte(e.Extra), nil
+		}
+	}
+
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	return nil, errors.New("key not found")
+}
+
+// Shamelessly copied from commands/routing.go
+func escapeDhtKey(s string) (string, error) {
+	parts := path.SplitList(s)
+	if len(parts) != 3 ||
+		parts[0] != "" ||
+		!(parts[1] == "ipns" || parts[1] == "pk") {
```

Comment:
I merged this, but I'm raising a question now: if it should be routing-system-agnostic, then it shouldn't be IPNS only right? Peer routing is still routing. This is for the implementation of the Routing API, and not for the IPNS record fetching directly, which happens in the gateway, through this API.

Reply:
You are right – I got confused by duplicated code, for some reason thought it's limited to our IPNS response  :see_no_evil:  

Only meant to remove `pk` from the context of IPNS response type on gateway.  
We should not change the behavior of `ipfs routing get` or generic RoutingAPI's  Get, and avoid duplicated code.

@hacdias mind cleaning this up? Since we are adding Routing.Get to CoreAPI, we should:

1. rename `ensureIPNSKey`  to `normalizeKey` (feels like a better description)
2. restore `|| parts[1] == "pk"`
3. switch `routing get` from `core/commands/routing.go` to use CoreAPI (so it looks similar to `blockGetCmd` in `core/commands/block.go`  and we have no duplicated routing code).